sygnal/helper/asyncio_connectprotocol.py

```diff
@@ -0,0 +1,178 @@
+
+    It performs the setup of a HTTP CONNECT proxy connection, then hands over
+    to another asyncio.Protocol.
+
+    For Twisted, see twisted_connectproxyclient.py instead.
+    """
+
+    def __init__(
+        self,
+        proxy_address: str,
+        target_address: str,
+        protocol_factory: Callable[[], Tuple[asyncio.Protocol, asyncio.Protocol]],
+        basic_proxy_auth: Optional[Tuple[str, str]],
+    ):
+        """
+        Args:
+            proxy_address (str):
+                The address of the HTTP proxy.
+                Used in the `Host` request header to the proxy.
+                Examples: '127.0.3.200:8080' or `prox:8080`
+            target_address (str):
+                The address of the destination that the proxy should connect to
+                on your behalf. Must include a port number.
+                Examples: 'example.org:443'
+            protocol_factory (() -> (Protocol, Protocol)):
+                A factory which should return a (bottom, top) pair of Protocols,
+                where the bottom and top are the bottom and top of a protocol
+                stack that will replace this HTTP Connect Proxy Protocol once
+                a tunnel is established.
+                (They may be the same if your protocol stack is 1-tall.)
+                Why? Well, you may wish to open a TLS-wrapped HTTP connection
+                in this proxy connection, which means your factory will need
+                to create a HttpProtocol (say) and an SSLProtocol to wrap it in.
+                The stack will change according to this diagram:
+
+                                                 +------------------+
+                                                 |  HTTP Protocol   |(a)
+                +---------------------+          +------------------+
+                | HTTP Proxy Protocol |   ===>   | SSL/TLS Protocol |(b)
+                +---------------------+----------+------------------+
+                |              Underlying TCP Transport             |
+                +---------------------------------------------------+
+
+                It's clear that this HTTP Proxy Protocol needs to know about the
+                SSL Protocol (b) so that it can switch the underlying transport's
+                protocol over.
+                However, the user of this class presumably wanted the HTTP
+                Protocol (a) for a reason, so we have to pass it back somehow
+                (via the `wait_for_establishment` Future in this case).
+
+                That's why the factory needs to return both the top and bottom
+                protocol of the stack (even if it is an unfortunately ugly
+                design).
+
+                Pseudo-example:
+                ```
+                def factory():
+                    http = HttpProtocol(...)
+                    ssl = SSLProtocol(app_protocol=http, ...)
+                    return (ssl, http)
+                ```
```

Comment:
yay for clear docs, but suggest taking some of this essay outside the `args` list!